core/packages/contracts/src/InboundChannel.sol

```diff
@@ -0,0 +1,155 @@
+    event VaultUpdated(address vault);
+    event RewardUpdated(uint256 reward);
+
+    error InvalidProof();
+    error InvalidNonce();
+    error InvalidHandler();
+    error NotEnoughGas();
+
+    constructor(IParachainClient _parachainClient, IVault _vault, uint256 _reward) {
+        _grantRole(ADMIN_ROLE, msg.sender);
+        parachainClient = _parachainClient;
+        vault = _vault;
+        reward = _reward;
+    }
+
+    function submit(
+        Message calldata message,
+        bytes32[] calldata leafProof,
+        bytes calldata headerProof
+    ) external {
+        bytes32 leafHash = keccak256(abi.encode(message));
+        bytes32 commitment = MerkleProof.processProof(leafProof, leafHash);
+        if (!parachainClient.verifyCommitment(commitment, headerProof)) {
+            revert InvalidProof();
+        }
+        if (message.nonce != nonce[message.origin] + 1) {
+            revert InvalidNonce();
+        }
+
+        // Increment nonce for origin.
+        // This ensures messages are not replayed. It also ensures re-entrancy protection,
+        // as a re-entrant call will be detected by the nonce check above.
+        //
+        // Sources of re-entrancy:
+        // * The relayer which gets forwarded ETH as a reward for submission
+        // * XCM::Transact calls to arbitrary untrusted contracts (not in scope for initial release)
+        nonce[message.origin]++;
+
+        // reward the relayer
+        // Should revert if there are not enough funds. In which case, the origin
+        // should top up the funds and have a relayer resend the message.
+        vault.withdraw(message.origin, payable(msg.sender), reward);
+
+        Handler memory handler = handlers[message.handler];
+        if (address(handler.recipient) == address(0)) {
+            revert InvalidHandler();
+        }
+
+        IRecipient recipient = IRecipient(handler.recipient);
+        uint256 gasToForward = uint256(handler.gasToForward);
+
+        // Ensure relayers pass enough gas for message to execute
+        if (gasleft() < gasToForward + GAS_BUFFER) {
+            revert NotEnoughGas();
+        }
+
+        DispatchResult memory result = DispatchResult(false, "", 0, hex"");
+
+        // Forward message to handler for execution
+        // Errors from the handler are ignored so as not to block the channel at the current nonce
+        try recipient.handle{ gas: gasToForward }(message.origin, message.payload) {
+            result.succeeded = true;
+        } catch Error(string memory reason) {
+            result.errorReason = reason;
+        } catch Panic(uint256 errorCode) {
+            result.errorPanicCode = errorCode;
+        } catch (bytes memory returnData) {
+            result.errorReturnData = returnData;
+        }
+
+        emit MessageDispatched(message.origin, message.nonce, result);
+    }
+
+    function updateHandler(uint16 id, Handler memory handler) external onlyRole(ADMIN_ROLE) {
```

Comment:
Handlers are permissioned here, only updatable by admin. I presume that's intentional for initial launch, and when we move to permissionless in future then this could be opened up?

Also, I can't see where this is actually called and where the handlers for the SovereignTreasury and NativeTokens are set? Is this done in the code/deployment at all?

Reply:
> Handlers are permissioned here, only updatable by admin. I presume that's intentional for initial launch, and when we move to permissionless in future then this could be opened up?

Nope! And this is very important. These handlers are simply components of the bridge and are within its trust domain. Equivalent to pallets in a substrate chain.

This narrowed design actually simplified a lot of things, for example around authorizing relationships between channels and handlers. You may remember previously we had that channel "operator" concept, where users would authorize apps to send messages on their behalf. Well all that complexity is now gone.

Over time we'll expand to include the following set of handlers:
* NativeTokens: Send Ethereum-native assets to Polkadot
* ForeignTokens: Send Polkadot-native assets to Ethereum
* SovereignTreasury: Accounts for sovereign parachains. Used to debit and credit fees/rewards for relayers
* Upgrades: Upgrade Support
* Transact: Support for XCM::Transact to arbitrary smart contracts.